### ISSUE TITLE
Shift right of negative signed ints is implementation defined in C

### ISSUE THREAD

**Issue by gambiteer** [gambit/gambit#218]
Shift right of negative signed ints is implementation defined in C.

We don't take that into account in the FIXASHR macro.

Do we care?

**Comment by feeley** [comment]
Yes we care... because there is also ```##fxwraplogical-shift-right``` when the width of the word affects the outcome. So someone would choose that primitive to avoid undefined behavior.

**Comment by gambiteer** [comment]
`##fxwraplogical-shift-right` doesn't do what we want, though:
```
> (##fxwraplogical-shift-right -1 1)
2305843009213693951
> (##fxwraplogical-shift-right -1 2)
1152921504606846975
> (fxarithmetic-shift-right -1 2)
-1
```
I think we may not care.

**Comment by feeley** [comment]
My point is that ```##fxwraplogical-shift-right``` has a well defined (but word-size dependent) semantics when the argument is negative. It would be useless to have ```##fxarithmetic-shift-right``` identically on non-negative, and have undefined semantics for negative. It is useful and consistent to have ```##fxarithmetic-shift-right``` extend the sign bit for negative numbers, because that is what the generic ```arithmetic-shift``` function does when shifting right (negative shift amount).

**Comment by gambiteer** [comment]
Perhaps we're talking past each other.

This part of the gcc manual

https://gcc.gnu.org/onlinedocs/gcc-6.3.0/gcc/Integers-implementation.html#Integers-implementation

says that right shift of negative integers does what we want.

This part of the LLVM manual

http://llvm.org/docs/LangRef.html#ashr-instruction

says that LLVM does what we want.

This part of the Microsoft Visual C manual

https://msdn.microsoft.com/en-us/library/aa243545(v=vs.60).aspx

says that Visual C does what we want.

Since the C standard says the behavior of right shift of negative numbers is "implementation defined" and these implementations define it the way we want, I think we should just drop the issue.

**Comment by feeley** [comment]
I see what you are saying. Ideally I'd like to have a ```#ifdef ___HAVE_GOOD_ASHR``` to select the fastest implementation of good right shift.
